### ISSUE TITLE
Determining type of JSON at runtume

### ISSUE THREAD

**Issue by themorlock** [stephenberry/glaze#173]
I had a question about determining what type of JSON is present at runtime.

I'm trying to parse data from an exchange and right now there's two different messages I can receive

a Pong:
`{"success":true,"ret_msg":"pong","conn_id":"daccac2a-c146-48ec-a069-90fd2bd7c48d","op":"ping"}`

or Market Data:
`{"topic":"orderbook.1.BTCUSDT","ts":1676942675027,"type":"delta","data":{"s":"BTCUSDT","b":[],"a":[["24840","0.76475"]],"u":39480281,"seq":3144408541}}`

Originally I figured I could use glz::get_sv_json or variant handling, but because neither of the responses have any common keys, trying to read the value of a key results in an error being thrown. Ideally, I don't want to use errors in my control flow, so I've resorted to modifying the raw string beforehand like this:

`data = data + "\"op\":\"update\""`

so when I parse data, I can check for the key 'op' and since 'op' already occurs in the Pong response I get different values and parse the JSON accordingly. However, I feel like this isn't ideal and I was wondering if there was potentially a faster and more elegant way to do this.

**Comment by mwalcott3** [comment]
Yeah variant type deduction support is pretty minimal right now and I could recommend some ways of handling your specific case but I prefer a more general solution and I think it is feasible in a relatively performant manner.

General solutions:
1. We need a better way of specifying custom type deductions
2. We could improve the auto deduction for the meta objects

One has been a longstanding issue that has been put off since we do not want to get locked into a bad interface and we haven't entirely decided what to do. Two can be implemented and improved later without breaking anyone using glaze since its internal so Id be more interested in tackling that issue right now.

I thought about it for alittle bit, and we have a lot of type info compile-time so we can deduce when certain combinations of keys have to result in one type with the variant. Here is a basic outline:
1. Create a compile time constexpr perfect hashmap of the key to uint64_t (or maybe a std::bitset or array of uint64_t to support more types in the variant but who uses more than 64 types in a variant) with each bit representing if the key belongs to the type for that type index in a variant. 
2. Start with a uint64_t possible_types with all bits set to one
3. Parse the key and lookup in the map and bitwise and the result with possible_types
4. Use std::has_single_bit to detect if a single type matches
5. If a single type matches parse as that type (Use one of the std bitscan functions to get the type index), if possible_types equals zero error out, otherwise skip to the next key (skip_ws match<':'> skip_ws, skip_value), and resume step 3. If we reach the end of the object we could either error out or parse it as the first type that matches. We might also want to keep track of the key and value string_views to speed up the parsing when we know the type (We know the max keys compile-time so we can put them in a std::array since dynamic memory allocation is expensive) but I think such optimizations can wait.

This would work in your case pretty well as there is no overlap in the keys so it could immediately deduce the type and would automatically select the correct type in a std::variant without any additional code on your end if implemented.

I'll talk to @stephenberry later to see what he thinks and try to brainstorm a bit but I think it is worth implementing and should be fairly straitforward.

**Comment by themorlock** [comment]
Amazing. After checking the response sizes for my situation I think I can deduce the type with straightforward `data.size()`comparison. Would be great to see those features in the future.

**Comment by mwalcott3** [comment]
@themorlock lol that's hilarious, sometimes the simplest solution is the best solution.

I think it's still a good idea to implement better auto deduction and I'll try to get that in as soon as the no_except branch is ironed out.

**Comment by mwalcott3** [comment]
Automatic object deduction in a std::variant based on unique key combinations was merged into the no_except branch in #175.
This also includes better tagged variant support. Still need to add some info on std::variant support and its limitations to the readme/docs.

no_except is being treated as the next major feature release branch (Some breaking changes are being made) and should be merged into main on this coming Monday so I'm closing this issue out.

Auto deduction example:
```c++
struct xy_t
{
   int x{};
   int y{};
};

template <>
struct glz::meta<xy_t>
{
   using T = xy_t;
   static constexpr auto value = object("x", &T::x, "y", &T::y);
};

struct yz_t
{
   int y{};
   int z{};
};

template <>
struct glz::meta<yz_t>
{
   using T = yz_t;
   static constexpr auto value = object("y", &T::y, "z", &T::z);
};

struct xz_t
{
   int x{};
   int z{};
};

template <>
struct glz::meta<xz_t>
{
   using T = xz_t;
   static constexpr auto value = object("x", &T::x, "z", &T::z);
};

suite metaobject_variant_auto_deduction = [] {
   "metaobject_variant_auto_deduction"_test = [] {
      std::variant<xy_t, yz_t, xz_t> var{};

      std::string b = R"({"y":1,"z":2})";
      expect(glz::read_json(var, b) == glz::error_code::none);
      expect(std::holds_alternative<yz_t>(var));
      expect(std::get<yz_t>(var).y == 1);
      expect(std::get<yz_t>(var).z == 2);

      b = R"({"x":5,"y":7})";
      expect(glz::read_json(var, b) == glz::error_code::none);
      expect(std::holds_alternative<xy_t>(var));
      expect(std::get<xy_t>(var).x == 5);
      expect(std::get<xy_t>(var).y == 7);

      b = R"({"z":3,"x":4})";
      expect(glz::read_json(var, b) == glz::error_code::none);
      expect(std::holds_alternative<xz_t>(var));
      expect(std::get<xz_t>(var).z == 3);
      expect(std::get<xz_t>(var).x == 4);
   };
};
```

Auto deduction works if the type can be determined by the combination of keys present. If you need to deduce the type based on the value associated with a key you can use the new tagged union support that now allows for custom tags instead of just "type" and the ids can be overridden from the default of glz::name_v of the type.
Tagged union types example:
```c++
struct put_action
{
   std::map<std::string, int> data{};
};

template <>
struct glz::meta<put_action>
{
   using T = put_action;
   static constexpr auto value = object("data", &T::data);
};

struct delete_action
{
   std::string data{};
};

template <>
struct glz::meta<delete_action>
{
   using T = delete_action;
   static constexpr auto value = object("data", &T::data);
};

using tagged_variant = std::variant<put_action, delete_action>;

template <>
struct glz::meta<tagged_variant>
{
   static constexpr std::string_view tag = "action";
   static constexpr auto ids = std::array{"PUT", "DELETE"}; //Defaults to glz::name_v of the type is ids is not supplied
};

suite tagged_variant_tests = [] {
   "tagged_variant_write_tests"_test = [] {
      // custom tagged discriminator ids
      tagged_variant var = delete_action{{"the_internet"}};
      std::string s{};
      glz::write_json(var, s);
      expect(s == R"({"action":"DELETE","data":"the_internet"})");
      s.clear();
   };
   
   "tagged_variant_read_tests"_test = [] {
      tagged_variant var{};
      expect(glz::read_json(var, R"({"action":"DELETE","data":"the_internet"})") == glz::error_code::none);
      expect(std::get<delete_action>(var).data == "the_internet");
   };
};
```

These two approaches should handle most use cases of determining the type of a json object at runtime. And they also work if the type of some item nested in a json object needs to be deduced. If someone runs into a real world case where this is insufficient they can open up a new issue and we will see what we can do. 

There are some things I could've added but Id rather not bloat the options so Ill wait to see if anyone actually stuff first before implementing it. Worst case we can add an option that loops through and trys all potential types in the variant to handle more complicated cases but due to the performance of such an approach I would never enable such a thing by default. Right now we error out if there are multiple types that match but we could also add an option for reading into the first matching type in the variant. Also right now the ids are limited to strings but they could easily be any value provided we can constexpr hash them, due to std::hash not being constexpr. In the short term we will most likely keep it to strings and supporting any type would require the user to supply a constexpr hashing function so Id prefer to wait for std::hash to become constexpr so we don't have to support user supplied hashing functions (Its easy to add things but hard to remove them).
